### DISCUSSION TITLE
How to find what node version a project supports?

### DISCUSSION THREAD

**Discussion by zainamin43** [nodejs/help#4757] [Q&A]
How to find what node version a project supports? When developer not specify in documentation OR package.json (engine)

**Comment by abhint** [comment]
You can specify the version of the node running stuff using package.json.
if you want more please visit: https://docs.npmjs.com/cli/v9/configuring-npm/package-json#engines

```json
{
  "engines": {
    "node": ">=0.10.3 <15"
  }
}
```

**Reply by zainamin43** [reply]
You don't understand my question. I am working on old project that does not specify a Node.js version in package.json. Like your reply. I have tried different Node.js versions, but they do not work.

**Reply by bnoordhuis** [reply]
I think you already know the answer: trial and error. You're probably hoping for an easier way out but there isn't one if your predecessor didn't write anything down.

**Reply by abhint** [reply]
> You don't understand my question. I am working on old project that does not specify a Node.js version in package.json. Like your reply. I have tried different Node.js versions, but they do not work.

Can you show the error?

**Reply by zainamin43** [reply]
> I think you already know the answer: trial and error. You're probably hoping for an easier way out but there isn't one if your predecessor didn't write anything down.

I wish

**Reply by zainamin43** [reply]
> > You don't understand my question. I am working on old project that does not specify a Node.js version in package.json. Like your reply. I have tried different Node.js versions, but they do not work.
> 
> Can you show the error?

{
  "name": "ABC Project",
  "version": "0.0.1",
  "author": "Ionic Framework",
  "homepage": "https://ionicframework.com/",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/common": "~10.2.3",
    "@angular/core": "~10.2.3",
    "@angular/forms": "~10.2.3",
    "@angular/platform-browser": "~10.2.3",
    "@angular/platform-browser-dynamic": "~10.2.3",
    "@angular/router": "~10.2.3",
    "@ionic-native/android-permissions": "^5.29.0",
    "@ionic-native/background-mode": "^5.29.0",
    "@ionic-native/camera": "^5.29.0",
    "@ionic-native/core": "^5.29.0",
    "@ionic-native/device": "^5.29.0",
    "@ionic-native/diagnostic": "^5.29.0",
    "@ionic-native/geolocation": "^5.29.0",
    "@ionic-native/http": "^5.29.0",
    "@ionic-native/location-accuracy": "^5.29.0",
    "@ionic-native/network": "^5.29.0",
    "@ionic-native/power-management": "^5.29.0",
    "@ionic-native/splash-screen": "^5.29.0",
    "@ionic-native/sqlite": "^5.29.0",
    "@ionic-native/sqlite-porter": "^5.29.0",
    "@ionic-native/status-bar": "^5.29.0",
    "@ionic/angular": "^5.4.4",
    "@ionic/storage": "^2.3.1",
    "cordova-plugin-advanced-http": "3.2.2",
    "cordova-plugin-android-permissions": "1.1.2",
    "cordova-plugin-file": "6.0.2",
    "cordova-plugin-geolocation": "4.0.2",
    "cordova-plugin-network-information": "3.0.0",
    "cordova-sqlite-storage": "6.0.0",
    "ion2-calendar": "^3.5.0",
    "moment": "^2.29.1",
    "recursive-readdir": "^2.2.2",
    "rxjs": "~6.6.3",
    "tslib": "^2.0.3",
    "zone.js": "~0.11.3"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.1002.0",
    "@angular/cli": "~10.2.0",
    "@angular/compiler": "~10.2.3",
    "@angular/compiler-cli": "~10.2.3",
    "@angular/language-service": "~10.2.3",
    "@ionic/angular-toolkit": "^2.3.3",
    "@types/jasmine": "~3.6.1",
    "@types/jasminewd2": "~2.0.8",
    "@types/node": "^14.14.7",
    "codelyzer": "^6.0.1",
    "cordova-android": "10.1.2",
    "cordova-plugin-background-mode": "^0.7.3",
    "cordova-plugin-camera": "6.0.0",
    "cordova-plugin-device": "^2.0.3",
    "cordova-plugin-ionic-keyboard": "^2.2.0",
    "cordova-plugin-powermanagement-orig": "^1.1.2",
    "cordova-plugin-request-location-accuracy": "^2.3.0",
    "cordova-plugin-splashscreen": "^6.0.0",
    "cordova-plugin-statusbar": "3.0.0",
    "cordova.plugins.diagnostic": "6.1.1",
    "jasmine-core": "~3.6.0",
    "jasmine-spec-reporter": "~6.0.0",
    "karma": "~5.2.3",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage-istanbul-reporter": "~3.0.3",
    "karma-jasmine": "~4.0.1",
    "karma-jasmine-html-reporter": "^1.5.4",
    "protractor": "~7.0.0",
    "ts-node": "~9.0.0",
    "tslint": "~6.1.0",
    "typescript": "~4.0.5",
    "uk.co.workingedge.cordova.plugin.sqliteporter": "^1.1.1"
  },
  "description": "An Ionic project",
  "cordova": {
    "plugins": {
      "cordova-plugin-whitelist": {},
      "cordova-plugin-statusbar": {},
      "cordova-plugin-device": {},
      "cordova-plugin-splashscreen": {},
      "cordova-plugin-ionic-webview": {
        "ANDROID_SUPPORT_ANNOTATIONS_VERSION": "27.+"
      },
      "cordova-plugin-ionic-keyboard": {},
      "cordova-plugin-background-mode": {},
      "at.gofg.sportscomputer.powermanagement": {},
      "uk.co.workingedge.cordova.plugin.sqliteporter": {},
      "cordova-plugin-request-location-accuracy": {
        "PLAY_SERVICES_LOCATION_VERSION": "16.+"
      },
      "cordova-plugin-camera": {
        "ANDROID_SUPPORT_V4_VERSION": "27.+"
      },
      "cordova-plugin-advanced-http": {
        "ANDROIDBLACKLISTSECURESOCKETPROTOCOLS": "SSLv3,TLSv1"
      },
      "cordova-plugin-network-information": {},
      "cordova-plugin-android-permissions": {},
      "cordova-plugin-geolocation": {},
      "cordova-sqlite-storage": {},
      "cordova.plugins.diagnostic": {
        "ANDROIDX_VERSION": "1.0.0",
        "ANDROIDX_APPCOMPAT_VERSION": "1.3.1"
      },
      "cordova-plugin-androidx": {},
      "cordova-plugin-androidx-adapter": {}
    },
    "platforms": [
      "android"
    ]
  }
}


How can find which node version is compatible from this?

**Comment by arpit9616** [comment]
If the project's required Node.js version isn't explicitly specified in the documentation or the `package.json` file, you can determine it by examining the GitHub release tags for all dependencies and inspecting their respective `package.json` files. This approach will allow you to derive a range of compatible Node.js versions, streamlining your analysis process.

For instance, let's consider the Angular framework as a central dependency. By referring to the specific version, such as `"@angular/core": "~10.2.3"`, you can explore its [dependency details](https://github.com/angular/angular/blob/10.2.3/package.json#L11C31-L11C31). This may lead you to infer a compatible Node.js version range, such as `">=10.9.0 <13.0.0"`, based on the Angular version's supported engine range.

Though this investigation might be somewhat time-consuming, particularly when applied to major dependencies, this method should help you gain valuable insights into the required Node.js environment. By repeating this process for the primary dependencies, you should be able to approximate the appropriate Node.js version for your project.

Hope this helps.
